server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -140,50 +138,131 @@ public CachingClusteredClient(
     );
   }
 
-  @Override
-  public Sequence<T> run(final QueryPlus<T> queryPlus, final Map<String, Object> responseContext)
+  private class SpecificQueryRunner
```

Comment:
This class is large enough to be extracted into a separate class. What do you think of separating it even though it is used only in ```CachingClusteredClient```? Also, a simple doc explaining what this class does would be good.